### PR TITLE
Fix NPE in JRTUtil

### PR DESCRIPTION
**Pull request by laeubi**
## What it does

In some cases JRTUtil#getJrtSystem() can return null what results in a NPE a caller most likley do not exspect.

This adds required null checks and act accordingly.

Fix https://github.com/eclipse-jdt/eclipse.jdt.core/issues/1154

## Author checklist

- [ ] I have thoroughly tested my changes
- [x] The change is following the [coding conventions](https://wiki.eclipse.org/Platform/How_to_Contribute#Coding_Conventions)
- [x] I have signed the [Eclipse Contributor Agreement (ECA)](https://www.eclipse.org/legal/ECA.php)

### CODE REVIEW

**Comment by laeubi** [comment]
@ jarthana @iloveeclipse can you take a look here, this is really annoying and currently makes working with the Java Search impossible in my Platform-IDE-SDK as it seems I have some project imported that triggers the NPE every time.
Test failure seems unrelated as I see it on other changes as well (e.g. here: https://ci.eclipse.org/jdt/job/eclipse.jdt.core-Github/job/PR-676/10/)

**Comment by iloveeclipse** [comment]
Please rebase to get rid of #1150 issue.

Also I believe your issue could be caused by https://github.com/eclipse-jdt/eclipse.jdt.debug/pull/231 and your fix is applied at wrong place.

I assume somehow the "autoimport" of JDK/JRE managed to setup invalid JDK/JRE.

Please check that: which is now the default one and how to prevent importing that one in first place.
Hint: I assume some project now uses JRE, not JDK as default, or something like this.

**Comment by laeubi** [comment]
> Also I believe your issue could be caused by https://github.com/eclipse-jdt/eclipse.jdt.debug/pull/231 and your fix is applied at wrong place.

I think this fix is good anyways, as obviously there can be `null` returned, so the code should handle this.

> I assume somehow the "autoimport" of JDK/JRE managed to setup invalid JDK/JRE.

I'll check that, how would I identify such "invalid JDK"? If i look at the dialog it all looks quite usual
grafik

I also selected each JDK and hit "edit" but no error was reported.

**Comment by iloveeclipse** [comment]
> I'll check that, how would I identify such "invalid JDK"?

If you can always reproduce your issue, try to debug it and see which JDK causes the problem.
Do you have other errors in the log that could give a hint?

**Comment by laeubi** [comment]
As it is on the batch-compiler it seems only system-out is printing something I started eclipse with a console now and see this error: https://github.com/eclipse-jdt/eclipse.jdt.core/issues/1154#issuecomment-1596509618

**Comment by laeubi** [comment]
@jarthana @iloveeclipse I have rebased this now and all tests seem green. If you think it should be fixed at another place I'm fine with that but this currently makes my Eclipse-Ide somewhat unuseable...

**Comment by iloveeclipse** [comment]
> If you think it should be fixed at another place I'm fine 

I think if possible we should check the presence of that file on adding such JRE's.

> with that but this currently makes my Eclipse-Ide somewhat unuseable...

Just remove or fix that Java 9 installation.

**Comment by laeubi** [comment]
> Just remove or fix that Java 9 installation (/usr/lib/jvm/java-9-openjdk-amd64/release is missing).

If you can point to the spec that it requires a `release` file to be present for a JVM I can try to "fix" this, but unless then I fear I won't get much attention for that, I probably can even create an empty file, remove the JVM and so on, but to the core I think JDT can simply better handle the case, a NPE is from my point of view never acceptable here and the code obviously **can** return null, so what ever seems appropriate (ignore, throw an documented exception, ...) should be applied to JDT in parallel.

**Comment by iloveeclipse** [comment]
> If you can point to the spec that it requires a `release` file to be present for a JVM I can try to "fix" this, but unless then I fear I won't get much attention for that

Well, almost all of JDK internal structure is not spec'd, even ct.sym is not - should we now rewrite all of JDT code that relies on the JDK internals? You've complained about, and I've offered a solution that would immediately help. You don't need to apply this solution. 

Note: the problem with "broken" JDK will appear surely later on in some different code in JDT, you may not even notice where it will be because the search or call hierarchy would return only a subset of matches.

**Comment by laeubi** [comment]
> Well, almost all of JDK internal structure is not spec'd

And that's why I suggest to handle such cases... some of the methods are even throwing an exception so callers can handle / forward this, but the NPE bubbles up into an ugly error dialog I have to dismiss all the time.

File: org.eclipse.jdt.core.compiler.batch/src/org/eclipse/jdt/internal/compiler/util/JRTUtil.java [hunk 1]
```diff
@@ -238,35 +239,67 @@ public static void reset() {
...
+		JrtFileSystem system = getJrtSystem(image, null);
+		if (system == null) {
+			return;
+		}
+		system.walkModuleImage(visitor, notify);
 	}
 
 	public static void walkModuleImage(File image, String release, final JRTUtil.JrtFileVisitor<java.nio.file.Path> visitor, int notify) throws IOException {
-		getJrtSystem(image, release).walkModuleImage(visitor, notify);
+		JrtFileSystem system = getJrtSystem(image, release);
```
Comment (iloveeclipse) [inline org.eclipse.jdt.core.compiler.batch/src/org/eclipse/jdt/internal/compiler/util/JRTUtil.java:250]: Ideally getJrtSystem() should re-throw original IO exception where possible, not simply return null or false (the caller is supposed to properly handle this case as exception).

I will push a follow up PR to cleanup this.

**Review by iloveeclipse — APPROVED** [review]
The change itself is not complete, but can be merged. I will push a follow up PR.

**Comment by iloveeclipse** [comment]
> I will push a follow up PR to cleanup this.

See https://github.com/eclipse-jdt/eclipse.jdt.core/pull/1170

**Comment by laeubi** [comment]
> deally getJrtSystem() should re-throw original IO exception where possible, not simply return null or false (the caller is supposed to properly handle this case as exception).

Thanks I also has thought about it, but was not sure about the consequences (especially as it seems one needs to enable exceptions explicitly). The followup looks like a good path forward!

**Pull request merged**